### PR TITLE
Implement views::counted

### PR DESCRIPTION
**Pull request by miscco**
Partially addresses #39

Note: Reading the requirement in http://eel.is/c++draft/ranges#range.counted-2 I was lazy and assumed that it is sufficient that the passed argument is convertible. Is that correct or do we need to add some `_Choice` magic?

### CODE REVIEW

File: stl/inc/ranges [hunk 1]
```diff
@@ -1913,6 +1913,44 @@ namespace ranges {
...
+                        noexcept(subrange{
+                            counted_iterator{_STD declval<_It>(), iter_difference_t<_It>{}}, default_sentinel})};
+                }
+            }
+
+            template <class _It>
+            static constexpr _Choice_t<_St> _Choice = _Choose<_It>();
+
+        public:
+            template <input_or_output_iterator _It>
```
Comment (miscco) [inline stl/inc/ranges:1940]: I am not really sure whether output iterators make sense here

Comment (miscco) [reply stl/inc/ranges:1940]: Scratch that, I checked and everything works with output iterators. I just had to improve the tests

**Comment by miscco** [comment]
@CaseyCarter, I am unsure whether we should test anything more as we essentially tested everything already with the span and subrange tests.

Other test ideas besides checking for content?

**Comment by miscco** [comment]
Urgh, that conversion warning of `auto result = ranges::views::counted(Iter{input}, 3);` sets me off.

I guess I should have been more careful regarding the `convertible_to<iter_difference<_Iter>>` requirement

**Comment by miscco** [comment]
@CaseyCarter I feel like the `convertible_to<_Diff, iter_difference_t<_It>>` requirement on `views::counted` would also make sense for  `counted_iterator`, where we explicitly require the second argument to be of type `iter_difference_t<_It>`

This was also the reason that in the test we have these two lines in the test:
```cpp
            counted_iterator<Iter> constructed(Iter{input}, iter_difference_t<Iter>{2});
            counted_iterator<Iter> constructedEmpty{Iter{input}, iter_difference_t<Iter>{0}};
```

Would you say that this is something we might want to change?

**Comment by miscco** [comment]
Also the actual error is that `std::span` takes an `size_t` as second argument, so I assume we would need to cast it to `size_t` before passing it so `std::span` even if that might be impossible :(

**Review by CaseyCarter — CHANGES_REQUESTED** [review]
Not a real review, just correcting the problem from #1393.

File: tests/std/tests/P0896R4_views_counted/test.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#include <concepts>
+#include <ranges>
+#include <span>
+
+#include <range_algorithm_support.hpp>
+
+using namespace std;
+
+template <input_or_output_iterator Iter>
+struct convertible_difference {
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_views_counted/test.cpp:15]: I suggest making this type non-movable or the hard-mode version of non-movable where SMFs `static_assert(always_false<Iter>)`.

File: tests/std/tests/P0896R4_views_counted/test.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <cassert>
+#include <concepts>
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_views_counted/test.cpp:6]: Unused, I think.

File: stl/inc/ranges [hunk 4]
```diff
@@ -1913,6 +1913,50 @@ namespace ranges {
...
+                } else if constexpr (_Choice<_It>._Strategy == _St::_Subrange) {
+                    return subrange{_First, _First + static_cast<iter_difference_t<_It>>(_Count)};
+                } else if constexpr (_Choice<_It>._Strategy == _St::_Subrange_counted) {
+                    return subrange{counted_iterator{_STD move(_First), static_cast<iter_difference_t<_It>>(_Count)},
+                        default_sentinel};
+                }
+            }
+        };
+
+        inline constexpr _Counted_fn counted;
```
Comment (CaseyCarter) [inline stl/inc/ranges:1959]: Could we please add this CPO to the `test_cpo` coverage in `P0896R4_ranges_range_machinery`?

File: stl/inc/ranges [hunk 5]
```diff
@@ -1913,6 +1913,50 @@ namespace ranges {
...
+            }
+
+            template <class _It>
+            static constexpr _Choice_t<_St> _Choice = _Choose<_It>();
+
+        public:
+            // clang-format off
+            template <input_or_output_iterator _It, class _Diff>
+                requires convertible_to<_Diff, iter_difference_t<_It>>
+            _NODISCARD constexpr auto operator()(_It _First, const _Diff _Count) const
```
Comment (CaseyCarter) [inline stl/inc/ranges:1943]: This (1) introduces an additional (user-observable) move for the iterator, and (2) requires the second argument to be copy constructible which the working draft does not. I suggest taking the first parameter by forwarding reference, and changing the type of the second parameter to the difference type of the first to force the conversion onto the caller. i.e., something like:
```suggestion
            template <class _It>
                requires input_or_output_iterator<decay_t<_It>>
            _NODISCARD constexpr auto operator()(
                _It&& _First, const iter_difference_t<decay_t<_It>> _Count) const
```
with fixups elsewhere to replace uses of `_It` with `decay_t<_It>` and perfect forward, etc.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/ranges [hunk 6]
```diff
@@ -1913,6 +1913,50 @@ namespace ranges {
...
+                // clang-format on
+                noexcept(_Choice<_It>._No_throw) {
+                _STL_VERIFY(_Count >= 0, "The size passed to views::counted must be non-negative");
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/ranges:1946]: Done

File: tests/std/tests/P0896R4_views_counted_death/test.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+#include <ranges>
+#include <vector>
+
+#include <test_death.hpp>
+using namespace std;
+
+static int some_ints[] = {0, 1, 2, 3};
+
+void test_constructor_negative_size() {
+    [[maybe_unused]] auto r = views::counted(some_ints, -3); // The size passed to views::counted must be non-negative
```
[the remark this answers is not in the document]
Comment (miscco) [reply tests/std/tests/P0896R4_views_counted_death/test.cpp:17]: :+1:

File: stl/inc/ranges (hunk 6 above)
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/ranges:1946]: Yes this should be an `_STL_ASSERT ` as this is a user facing check and not some internal verification

File: tests/std/tests/P0896R4_views_counted/test.cpp (hunk 2 above)
Comment (miscco) [reply tests/std/tests/P0896R4_views_counted/test.cpp:15]: deleted the move SMFs

File: tests/std/tests/P0896R4_views_counted/test.cpp (hunk 3 above)
Comment (miscco) [reply tests/std/tests/P0896R4_views_counted/test.cpp:6]: Removed

File: stl/inc/ranges (hunk 4 above)
Comment (miscco) [reply stl/inc/ranges:1959]: Done

File: tests/std/tests/P0896R4_views_counted_death/test.cpp [hunk 8]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#define _CONTAINER_DEBUG_LEVEL 1
+
+#include <cassert>
+#include <cstddef>
+#include <ranges>
+#include <vector>
```
[the remark this answers is not in the document]
Comment (miscco) [reply tests/std/tests/P0896R4_views_counted_death/test.cpp]: :+1:

File: stl/inc/ranges [hunk 9]
```diff
@@ -1913,6 +1913,48 @@ namespace ranges {
...
 
         inline constexpr _Drop_fn drop;
+
+        // VARIABLE views::counted
+        class _Counted_fn {
+        private:
+            enum class _St { _Span, _Subrange, _Subrange_counted };
+
+            template <class _It>
+            _NODISCARD static _CONSTEVAL _Choice_t<_St> _Choose() noexcept {
```
Comment (CaseyCarter) [inline stl/inc/ranges:1923]:
```suggestion
            _NODISCARD static _CONSTEVAL _Choice_t<_St> _Choose() noexcept {
                _STL_INTERNAL_STATIC_ASSERT(input_or_output_iterator<_It>);
```

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/ranges [hunk 10]
```diff
@@ -1913,6 +1913,49 @@ namespace ranges {
...
+                requires input_or_output_iterator<decay_t<_It>>
+            _NODISCARD constexpr auto operator()(_It&& _First, const iter_difference_t<decay_t<_It>> _Count) const
+                noexcept(_Choice<decay_t<_It>>._No_throw) {
+                // clang-format on
+                _STL_ASSERT(_Count >= 0, "The size passed to views::counted must be non-negative");
+                if constexpr (_Choice<decay_t<_It>>._Strategy == _St::_Span) {
+                    return span{_STD to_address(_STD forward<_It>(_First)), static_cast<size_t>(_Count)};
+                } else if constexpr (_Choice<decay_t<_It>>._Strategy == _St::_Subrange) {
+                    return subrange{_First, _First + _Count};
+                } else if constexpr (_Choice<decay_t<_It>>._Strategy == _St::_Subrange_counted) {
```
Comment (StephanTLavavej) [inline stl/inc/ranges:1952]: Suggestion: This repeats `_Choice<decay_t<_It>>._Strategy` 3 times. In the Spaceship CPOs, I used `constexpr _St _Strat = MEOW` to make this more readable.

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/ranges (hunk 10 above)
Comment (miscco) [reply stl/inc/ranges:1952]: Yeah, I was thinking about that one too :+1:

Comment (StephanTLavavej) [reply stl/inc/ranges:1952]: I'll validate and push a change so we can move this to Ready To Meow.

Comment (StephanTLavavej) [reply stl/inc/ranges:1952]: Actually, I will file an issue for a global cleanup. There are lots of occurrences in `<ranges>` and `<xutility>` that should be changed consistently.

Comment (StephanTLavavej) [reply stl/inc/ranges:1952]: Filed #1419.

**Comment by CaseyCarter** [comment]
Every contribution counts!

**Pull request merged**